components/sup/src/logger/env_logger_compatibility.rs

```diff
@@ -0,0 +1,246 @@
+//! Transition code to bridge users from `env_logger` to `log4rs`.
+//!
+//! Takes a configuration string (e.g., the content of the environment
+//! variable `RUST_LOG`) for configuring `env_logger` and transforms
+//! it (as much as possible) into a valid `log4rs` configuration.
+//!
+//! An important difference is in how module targeting works. With
+//! `env_logger`, you could do something like this:
+//!
+//! ```text
+//! RUST_LOG=debug,tokio=error
+//! ```
+//!
+//! This would ensure that all `debug` and higher severity messages
+//! were logged, but every module path that started with the string
+//! "tokio" would only log for `error` and higher (Tokio can be rather
+//! verbose at the lower logging levels).
+//!
+//! The `log4rs` library treats this target slightly differently; it
+//! *must* be a Rust module path. That is, the same "tokio=error"
+//! targeting string would match the module `tokio`, as well as
+//! `tokio::foo`, `tokio::foo::bar`, etc. However, for this particular
+//! example, there isn't actually a `tokio` module at the root;
+//! instead you have things like `tokio_reactor`, `tokio_threadpool`,
+//! etc. In `log4rs`, "tokio" will _not_ match `tokio_reactor` or
+//! `tokio_threadpool`; thus, you would see debug-level messages from
+//! these modules! To filter these in `log4rs`, you would need to
+//! explicitly list the module prefixes, like so:
+//!
+//! ```text
+//! RUST_LOG=debug,tokio_reactor=error,tokio_threadpool=error
+//! ```
+//!
+//! In other words, `log4rs` is much stricter in its notion of
+//! what qualifies as a module prefix (indeed, it actually _is_ a
+//! module prefix, rather than a simple string prefix).
+//!
+//! Another notable difference is that while `env_logger` permits the
+//! addition of a regular expression filter (e.g., only print messages
+//! that match the expression "foo*"), we do *not*. Any filter that is
+//! present will be ignored.
+//!
+//! We also don't honor the `RUST_LOG_STYLE` variable; no output is
+//! colored at all.
+//!
+//! See https://docs.rs/env_logger/0.6.1/env_logger/ for more details.
+use log::LevelFilter;
+use log4rs::{append::console::ConsoleAppender,
+             config::{Appender,
+                      Config,
+                      Logger,
+                      Root},
+             encode::pattern::PatternEncoder};
+use std::{collections::HashMap,
+          str::FromStr};
+
+pub(super) fn from_env() -> Option<Config> {
+    std::env::var("RUST_LOG").ok().map(|env_var| {
+                                      eprintln!("RUST_LOG environment variable found; using it \
+                                                 to configure log4rs");
+                                      env_var.parse::<EnvLogConfig>()
+                                             .expect("RUST_LOG parsing can't fail")
+                                             .into()
```

Comment:
Since it's infallible, why not implement `From<String> for EnvLogConfig` instead of `FromStr` and simplify this to:
```rust
EnvLogConfig::from(env_var).into()
```
?

Reply:
I'd rather leave this as `FromStr`, even though it is infallible, since this is inherently a parsing activity.

This question did cause me to dig in and figure out exactly what the distinction between `FromStr` and `From<String>` actually is, though; I actually wasn't too clear on that myself! After looking at the `From<String>` implementations in the standard library, it looks like `From<String>` is just for doing things like creating contained strings (or string-like things), as in `From<String> for Box<String>`, or for converting from a string into another inherently string-like thing, as with `From<String> for PathBuf`. 

That feels like a very distinct use case from what we're trying to do here.

Reply:
I mostly suggested it because it eliminated the need for both a turbo fish and an `expect`, but if you feel `FromStr` is more appropriate, that's fine with me.